src/fragments/lib/graphqlapi/js/optimistic-ui.mdx

```diff
@@ -0,0 +1,1049 @@
+
+<Callout>
+
+TanStack Query Devtools are not required, but are a useful resource for debugging and understanding how TanStack works under the hood. By default, React Query Devtools are only included in bundles when `process.env.NODE_ENV === 'development'`, meaning that no additional configuration is required to exclude them from a production build. 
+For more information on the TanStack Query Devtools, visit the [TanStack Query Devtools docs](https://tanstack.com/query/latest/docs/react/devtools)
+
+</Callout>
+
+```ts
+import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
+import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
+
+// Create a client
+const queryClient = new QueryClient()
+```
+
+Next, provide the client to your app:
+
+```ts
+<QueryClientProvider client={queryClient}>
+  <App />
+  <ReactQueryDevtools initialIsOpen={false} />
+</QueryClientProvider>
+```
+
+<Callout>
+
+For the complete working example, including required imports and React component state management, see the [Complete Example](#complete-example) below.
+
+</Callout>
+
+## How to use TanStack Query query keys with the GraphQL API
+
+TanStack Query manages query caching based on query keys. Query keys have to be an array at the top level, and can be as simple as an array with a single string, or as complex as an array of many strings and nested objects. The query key must be serializable, and unique to the query's data. 
+
+When using TanStack to render optimistic UI with the GraphQL API, it is important to note that different query keys must be used depending on the API operation. When retrieving a list of items, a single string is used (e.g. `queryKey: ["realEstateProperties"]`). This query key is also used to optimistically render a newly created item. When updating or deleting an item, the query key must also include the unique identifier for the record being deleted or updated (e.g. `queryKey: ["realEstateProperties", newRealEstateProperty.id]`).
+
+For more detailed information on query keys, see the [TanStack Query documentation](https://tanstack.com/query/v4/docs/react/guides/query-keys).
+
+## Optimistically rendering a list of records
+
+To optimistically render a list of items returned from the GraphQL API, use the TanStack `useQuery` hook, passing in the GraphQL API query as the `queryFn` parameter. The following example creates a query to retrieve all records from the API. We'll use `realEstateProperties` as the query key, which will be the same key we use to optimistically render a newly created item.
+
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+const {
+  data: realEstateProperties,
+  isLoading,
+  isSuccess,
+  isError: isErrorQuery,
+} = useQuery({
+  queryKey: ["realEstateProperties"],
+  // TODO: improve return type
+  queryFn: async (): Promise<
+    (RealEstateProperty | null | undefined)[] | null | undefined
```

Comment:
If you omit this type completely does the return type of `realEstateProperties` not get auto-inferred? (Same question applies to all `queryFn` return types later in the doc - can we avoid adding all of those?)

Reply:
Good callout - the TanStack docs recommended typing the return type of the `queryFn`, so I did it and planned to revisit. Auto-inference works better here! Updated :)